src/newrelic/collector/client.go

```diff
@@ -6,367 +6,335 @@
+}
+// IsRestartException indicates that the agent should restart.
+func (resp RPMResponse) IsRestartException() bool {
+    return resp.StatusCode == 401 || resp.StatusCode == 409
+}
+// ShouldSaveHarvestData indicates that the agent should save the data and try
+// to send it in the next harvest.
+func (resp RPMResponse) ShouldSaveHarvestData() bool {
+    switch resp.StatusCode {
+    case 408, 429, 500, 503:
+        return true
+    default:
+        return false
+    }
 }
 
 // The agent languages we give the collector are not necessarily the ideal
 // choices for a user agent string.
 var userAgentMappings = map[string]string{
-	"c":   "C",
-	"sdk": "C",
-	"php": "PHP",
-	"":    "Native",
+    "c":   "C",
+    "sdk": "C",
+    "php": "PHP",
+    "":    "Native",
 }
 
-func (cmd *Cmd) userAgent() string {
-	if cmd.ua == "" {
-		lang := cmd.AgentLanguage
-		if val, ok := userAgentMappings[cmd.AgentLanguage]; ok {
-			lang = val
-		}
+func (control *RpmControls) userAgent() string {
+    if control.ua == "" {
+        lang := control.AgentLanguage
+        if val, ok := userAgentMappings[control.AgentLanguage]; ok {
+            lang = val
+        }
 
-		ver := "unknown"
-		if cmd.AgentVersion != "" {
-			ver = cmd.AgentVersion
-		}
+        ver := "unknown"
+        if control.AgentVersion != "" {
+            ver = control.AgentVersion
+        }
 
-		cmd.ua = fmt.Sprintf("NewRelic-%s-Agent/%s NewRelic-GoDaemon/%s", lang,
-			ver, version.Number)
-	}
+        control.ua = fmt.Sprintf("NewRelic-%s-Agent/%s NewRelic-GoDaemon/%s", lang,
+                                 ver, version.Number)
+    }
 
-	return cmd.ua
+    return control.ua
 }
 
 type Client interface {
-	Execute(cmd Cmd) ([]byte, error)
+    Execute(cmd RpmCmd, cs RpmControls) RPMResponse
 }
 
-type ClientFn func(cmd Cmd) ([]byte, error)
+type ClientFn func(cmd RpmCmd, cs RpmControls) RPMResponse
 
-func (fn ClientFn) Execute(cmd Cmd) ([]byte, error) {
-	return fn(cmd)
+func (fn ClientFn) Execute(cmd RpmCmd, cs RpmControls) RPMResponse {
+    return fn(cmd, cs)
 }
 
 type limitClient struct {
-	timeout   time.Duration
-	orig      Client
-	semaphore chan bool
+    timeout   time.Duration
+    orig      Client
+    semaphore chan bool
 }
 
-func (l *limitClient) Execute(cmd Cmd) ([]byte, error) {
-	var timer <-chan time.Time
-
-	if 0 != l.timeout {
-		timer = time.After(l.timeout)
-	}
-
-	select {
-	case <-l.semaphore:
-		defer func() { l.semaphore <- true }()
-		b, err := l.orig.Execute(cmd)
-		return b, err
-	case <-timer:
-		return nil, fmt.Errorf("timeout after %v", l.timeout)
-	}
+func (l *limitClient) Execute(cmd RpmCmd, cs RpmControls) RPMResponse {
+    var timer <-chan time.Time
+
+    if 0 != l.timeout {
+        timer = time.After(l.timeout)
+    }
+
+    select {
+    case <-l.semaphore:
+        defer func() { l.semaphore <- true }()
+        resp := l.orig.Execute(cmd, cs)
+        return resp
+    case <-timer:
+        return RPMResponse{}
```

Comment:
Does the need to set the ```Err``` field on the case of a timeout?  It seems the code in ```processor.go``` which calls this will check the ```Err``` field and if it is not set then it will act as if the function succeeded instead of timeing out as was the case.